src/helpers/buffers/unordered_receiver.rs

```diff
@@ -204,29 +209,54 @@ where
     /// Poll for the next record.  This should only be invoked when
     /// the future for the next message is polled.
     fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
-        if let Some(m) = self.spare.read() {
-            self.wake_next();
-            return Poll::Ready(Ok(m));
-        }
-
-        loop {
-            match self.stream.as_mut().poll_next(cx) {
-                Poll::Pending => {
-                    return Poll::Pending;
-                }
-                Poll::Ready(Some(b)) => {
-                    if let Some(m) = self.spare.extend(b.as_ref()) {
-                        self.wake_next();
-                        return Poll::Ready(Ok(m));
+        // If spare has enough data for us, poll it first
+        // otherwise, poll the underlying stream until it returns pending or it provides enough
+        // data to return a value.
+
+        let message = self.spare.read::<M>();
+        let next = match message {
+            Some(m) => {
+                // this check exists to make sure the inner stream is eventually moved to
+                // the closed state. We don't want to poll it too often, but we also need to know
+                // when it is done and `UnorderedReceiver` can be dropped.
+                if self.spare.is_empty() {
+                    // we don't want to be woken up here, control loop is driven by the client.
+                    // They decide when they want the next message and must issue a `poll` for it.
+
+                    // TODO: https://github.com/rust-lang/rust/issues/98286
+                    let mut cx = Context::from_waker(futures::task::noop_waker_ref());
+                    match self.stream.as_mut().poll_next(&mut cx) {
+                        Poll::Ready(Some(bytes)) => {
+                            // Spare is empty because of the check above.
+                            self.spare.replace(bytes.as_ref());
+                        }
+                        Poll::Ready(None) | Poll::Pending => {}
                     }
                 }
-                Poll::Ready(None) => {
-                    return Poll::Ready(Err(Error::EndOfStream {
-                        record_id: RecordId::from(self.next),
-                    }));
-                }
+
+                Poll::Ready(Ok(m))
             }
+            None => loop {
+                match ready!(self.stream.as_mut().poll_next(cx)) {
+                    Some(bytes) => {
+                        if let Some(m) = self.spare.extend(bytes.as_ref()) {
+                            break Poll::Ready(Ok(m));
+                        }
+                    }
+                    None => {
+                        break Poll::Ready(Err(Error::EndOfStream {
```

Comment:
You should flush `self.spare` if you hit this arm, see below.

Reply:
Now I am convinced that `M` must be on `OperatingState` rather than on `poll_next`. Will make this change

Reply:
I take this back - moving `M` up is not possible because we need heterogeneous channels and we keep them all in the same map. 

I will still close the receiver, but will add a note to the `poll_next` as this behavior may be confusing in cases when we read different messages off the same channel